spec/SPEC-annex-C-source-embedding.md

```diff
@@ -0,0 +1,77 @@
+## Annex C
+
+Annex B contains a method of embedding an Input Manifest Identifier into source code files files.
+
+### Embedded Input Manifest Identifier
+
+Most source code files are hand coded by humans.  Some however are generated from other input(s) by a build tool.
+
+A build tool outputing a source code file may embed the Input Manifest Identifier for the output source code file into
+the output source code file by adding a comment line containing a string of the form:
+
+``` OmniBOR-Input-Manifest-ID: [ ${comma separated list of Input Manifest Identifier URIs} ]
+```
+
+For a file with C commenting semantics (like C, C++, Java, Go, etc) a concrete example might be:
+
+```// OmniBOR-Input-Manifest-ID: [ gitoid:blob:sha1:261eeb9e9f8b2b4b0d119366dda99c6fd7d35c64, gitoid:blob:sha256:09c825ac02df9150e4f93d12ba1da5d1ff5846c3e62503c814aa3a300c535772 ]
+```
+
+For a file with shell commenting semantics (like a shell script, Python, etc) a concrete example might be:
+```# OmniBOR-Input-Manifest-ID: [ gitoid:blob:sha1:261eeb9e9f8b2b4b0d119366dda99c6fd7d35c64, gitoid:blob:sha256:09c825ac02df9150e4f93d12ba1da5d1ff5846c3e62503c814aa3a300c535772 ]
+```
+
+### Placement of OmniBOR-Input-Manifest-ID Comment Line
+
+The OmniBOR-Input-Manifest-ID comment line should be placed before the first non-comment, non-blank text in the file.  The OmniBOR-Input-Manifest-ID comment line should be proceeded and followed by a blank line to insure it is not interpretted as part of another comment block.
+
+A subsequent build tool reading the source code file should interpret the first such comment line it encounters in the file as being the Input Manifest Identifier, and ignore subsequent comment lines in the file which may contain Input Manifest Identifiers.
```

Comment:
```suggestion
A subsequent build tool reading the source code file should interpret the last such comment line it encounters in the file as being the Input Manifest Identifier, and ignore previous comment lines in the file which may contain Input Manifest Identifiers.
```

Reply:
Alternate proposal that may be easier for implementers:

- Append OmniBOR comment line to the end of the file.

Reply:
We should probably also make sure our tooling provides an easy way to answer the question: what is the `OmniBOR-Input-Manifest-ID` of this file, to enable debugging / assumption checking for people using OmniBOR in practice (meaning: you can check that it's concluding the ID is the one you expect).

Reply:
@alilleybrinker Could you say more... how are you envisioning that the ID is the one you expect?

Reply:
Okay, imagining we have an `omnibor` CLI, it would be nice to be able to do something like (ignore the specifics of the API design implied here):

```sh
$ omnibor manifest-id <file>
```

Which would print the "winning" manifest ID from the file. You could also, in theory, do:

```sh
$ omnibor manifest-id --all <file>
```

Which would print the full listing of ID's found in the file, in order. Maybe even also have:

```sh
$ omnibor manifest-id --all --line-numbers <file>
```

Which would print the same as `--all`, but with the line number of each ID prepended.

---

With these above commands existing, users could then debug surprising OmniBOR behavior (for example, the artifact ID's of two things not being equal when they believe they should be).